picard/oauth.py

```diff
@@ -244,14 +247,24 @@ def fetch_username(self, callback):
 
     def on_fetch_username_finished(self, callback, data, http, error):
         successful = False
+        error_msg = None
         try:
             if error:
                 log.error("OAuth: username fetching failed: %s", data)
+                error_msg = self._extract_error_description(http, data)
             else:
                 self.username = data["sub"]
                 log.debug("OAuth: got username %s", self.username)
                 successful = True
         except Exception as e:
             log.error('OAuth: Unexpected error handling username fetch response: %r', e)
+            error_msg = _('Unexpected authentication error')
         finally:
-            callback(successful)
+            callback(successful=successful, error_msg=error_msg)
+
+    def _extract_error_description(self, http, data):
+        if self.webservice.http_response_code(http) == 400:
```

Comment:
Shouldn't we handle any error code here?

Reply:
In that case we have a defined behavior and know how to extract the error message.

What would you do for other cases? Return a generic message, maybe with the HTTP code?

Reply:
I was just thinking about something like:

 ```python
http_error_code = self.webservice.http_response_code(http)
if http_error_code >= 400:
    response = load_json(data)
    if 'error_description' in response:
        return response['error_description'] + " (%d)" % http_error_code
```

Reply:
In practice this would just fail completely in most cases except the expected 400, the else would most likely not be hit. That's because as long as the endpoint will be reached and reacts the only error it will return is a 400. The other cases are more serious, such as a 5xx code by the webserver, or maybe a generic 404 error. Those likely won't parse as JSON.

I did it a bit different in c03716a25b752365f5f0c206b5982246130c326f so it shows a fallback message whenever parsing the JSON or reading `error_description`  fails. I also wouldn't expose the HTTP code in the 400 case to the UI. That's already part of the webserver module's logging.